File: week3/course_skeleton/course_skeleton.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

typedef struct student {
  //TODO
} Student;

typedef struct course {
  Student* head;
} Course; 

bool isEmpty(Course* c) {
  return (c->head == NULL);
}

bool addStudent(Course* c, int id, char* name) {
  //TODO
}

bool deleteStudent(Course* c, int id) {
  //TODO
}

Student* find(Course* c, int id) {
  //TODO
}

void write(Course* c, FILE* outFile) {
  //TODO
}

int main(int argc, char* argv[]) {
  if (argc != 3) {
    printf("Correct usage: [program] [input] [output]\n");
    exit(1);
  }
  FILE* inFile = fopen(argv[1], "r");
  if (inFile == NULL) {
    printf("no such file %s\n", argv[1]);
    exit(1);
  }
  FILE* outFile = fopen(argv[2], "w");
  if (outFile == NULL) {
    printf("no such file %s\n", argv[2]);
    exit(1);
  }
  Course course;
  char line[512];
  while (fgets(line, sizeof(line), inFile))
  {
    char* token = strtok(line, " \n");
    const char op = line[0];
    int id;
    Student* found = NULL;
    token = strtok(NULL, " \n");
    if (!token) {
      printf("invalid input\n");
      exit(1);
    }
    id = atoi(token);
    switch(op)
    {
      case 'A':
        token = strtok(NULL, " \n");
        if (!token) {
          printf("ADD: invalid input\n");
          exit(1);
        }
        if (addStudent(&course, id, token))
          write(&course, outFile);
        else
          fprintf(outFile, "Addition failed\n");
        break;
      case 'D':
        if (deleteStudent(&course, id))
          write(&course, outFile);
        else
          fprintf(outFile, "Deletion failed\n");
        break;
      case 'F':
        found = find(&course, id);
        if (found == NULL)
          fprintf(outFile, "Search failed\n");
        else
          fprintf(outFile, "%d %s\n", found->id, found->name);
        break;
      default:
        printf("Undefined operator\n");
        exit(1);
    }
  }
  fclose(outFile);
  fclose(inFile);
}

File: week3/course_skeleton/course_skeleton.py
import sys
ADD = 'A'
DELETE = 'D'
FIND = 'F'

class Student:
  def __init__(self, i, n, p=None):
    #TODO
    pass

class Course:
  def __init__(self):
    #TODO
    pass

  def __len__(self):
    return self.size

  def isEmpty(self):
    return self.size == 0

  def addStudent(self, newID, newName):
    #TODO
    pass

  def deleteStudent(self, queryID):
    #TODO
    pass

  def find(self, queryID):
    #TODO
    pass

  def write(self, outFile):
    #TODO
    pass

if __name__ == "__main__":
  if len(sys.argv) != 3:
    raise Exception("Correct usage: [program] [input] [output]")
  
  course = Course()
  with open(sys.argv[1], 'r') as inFile:
    lines = inFile.readlines()
  with open(sys.argv[2], 'w') as outFile:
    for line in lines:
      words = line.split()
      op = words[0]
      if op == ADD:
        if len(words) != 3:
          raise Exception("ADD: invalid input")
        i, n = int(words[1]), words[2]
        if course.addStudent(i, n):
          course.write(outFile)
        else:
          outFile.write("Addition failed\n")
      elif op == DELETE:
        if len(words) != 2:
          raise Exception("DELETE: invalid input")
        i = int(words[1])
        if course.deleteStudent(i):
          course.write(outFile)
        else:
          outFile.write("Deletion failed\n")
      elif op == FIND:
        if len(words) != 2:
          raise Exception("DELETE: invalid input")
        i = int(words[1])
        found = course.find(i)
        if not found:
          outFile.write("Search failed\n")
        else:
          outFile.write(str(found.id) + " " + found.name + "\n")
      else:
        raise Exception("Undefined operator")
        
